### PR TITLE
Add proxy for TinyOWS

### PR DESCRIPTION
**Pull request by tsauerwein**
This PR adds a proxy for TinyOWS which checks if a user has the rights to access a layer. A user has the right to read or write a layer, if the layer belongs to one of the user's restrictions areas and the `readwrite` flag of that restriction area is set to `True`.

It is important to note that the proxy does not take into account the area of a restriction area, contrary to the MapServer proxy which makes sure that only the features inside the defined area are returned. This is because the restriction extent would have to be hard-coded in the TinyOWS configuration file (see attribute [geobbox](http://mapserver.org/tinyows/configfile.html#layer-element)). So, if the access should restricted to a certain area, this area would have to be set in the configuration file.

This PR is not yet finished, but reviews are already welcome! Still to do:

* [x] Update documentation
* [x] Update demo

### CODE REVIEW

**Comment by sbrunner** [comment]
Looks good, when documentation is done and Travis tests pass you can merge :-)

File: c2cgeoportal/views/tinyowsproxy.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+        Checks if the current user has the rights to access the given
+        type-names.
+        """
+        writable_layers = get_writable_layers(self.role_id)
+        return typenames.issubset(writable_layers)
+
+    def _get_headers(self):
+        headers = Proxy._get_headers(self)
+        if self.settings.get("tinyows_host", "") != "":
+            headers["Host"] = self.settings.get("tinyows_host")
```
Comment (sbrunner) [inline c2cgeoportal/views/tinyowsproxy.py]: I don't understand why you need this ...

File: c2cgeoportal/views/tinyowsproxy.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+        """
+        Checks if the current user has the rights to access the given
+        type-names.
+        """
+        writable_layers = get_writable_layers(self.role_id)
+        return typenames.issubset(writable_layers)
+
+    def _get_headers(self):
+        headers = Proxy._get_headers(self)
+        if self.settings.get("tinyows_host", "") != "":
```
Comment (sbrunner) [inline c2cgeoportal/views/tinyowsproxy.py]: `if "tinyows_host" in self.settings:`
isn't better ?

File: c2cgeoportal/views/tinyowsproxy.py (hunk 1 above)
Comment (tsauerwein) [inline c2cgeoportal/views/tinyowsproxy.py]: This is only for the debug mode (`make serve`).

File: c2cgeoportal/views/tinyowsproxy.py (hunk 2 above)
Comment (tsauerwein) [inline c2cgeoportal/views/tinyowsproxy.py]: `self.settings.get("tinyows_host", "") != ""` checks two things: If there is an entry with the key `tinyows_host` and also if a possible value is a non-empty string. So it's not the same as `"tinyows_host" in self.settings`.

**Comment by tsauerwein** [comment]
I updated the documentation and demo. Please review.

ping @yjacolin

File: c2cgeoportal/views/tinyowsproxy.py (hunk 2 above)
Comment (sbrunner) [inline c2cgeoportal/views/tinyowsproxy.py]: Yes but settings `tinyows_host` to "" should be en error not a disable ...

**Comment by sbrunner** [comment]
two untested lines:
https://coveralls.io/builds/2393880/source?filename=c2cgeoportal%2Fviews%2Ftinyowsproxy.py#L134
https://coveralls.io/builds/2393880/source?filename=c2cgeoportal%2Fviews%2Ftinyowsproxy.py#L164
And I should know why you requires this `tinyows_host`

File: c2cgeoportal/views/tinyowsproxy.py (hunk 2 above)
Comment (tsauerwein) [inline c2cgeoportal/views/tinyowsproxy.py]: `tinyows_host` is optional. If it's not given in the settings, it's not set on the header.

Comment (sbrunner) [inline c2cgeoportal/views/tinyowsproxy.py]: Then `tinyows_host` shouldn't be in the settings ...

Comment (tsauerwein) [inline c2cgeoportal/views/tinyowsproxy.py]: Where else?

I know you are not using `make serve` but others are. And when you are using `make serve`, you explicitly have to set the host. Then it would be consistent to set the host together with all the other settings for TinyOWS. And like it is [documented](https://github.com/tsauerwein/c2cgeoportal/blob/73484ab87d6061013dde5bd980b15ec25205cfd2/c2cgeoportal/scaffolds/update/CONST_vars.yaml_tmpl#L186-189), it is not required to set the host.

Comment (sbrunner) [inline c2cgeoportal/views/tinyowsproxy.py]: Thanks for the second response, but set `tinyows_host` to "" steal haven't any sense ...

**Comment by sbrunner** [comment]
Thanks for the tests :-)

File: c2cgeoportal/views/tinyowsproxy.py (hunk 2 above)
Comment (tsauerwein) [inline c2cgeoportal/views/tinyowsproxy.py]: It is not set to `""`.

    if self.settings.get("tinyows_host", "") != "":

is the same as:

    if self.settings.get("tinyows_host") is not None and self.settings.get("tinyows_host") != "":

Comment (sbrunner) [inline c2cgeoportal/views/tinyowsproxy.py]: Yes I know it but why testing that `tinyows_host` is not `""` is un nessery ... it's yhy I largely prefer:
`if "tinyows_host" in self.settings:`

Comment (tsauerwein) [inline c2cgeoportal/views/tinyowsproxy.py]: Why is it unnecessary? I only want to change the host if a valid host name is given. That is a non-empty string.

Because this is only used for debugging, it's not so important. But I don't see your point.

Comment (sbrunner) [inline c2cgeoportal/views/tinyowsproxy.py]: More readable syntax ...

Comment (tsauerwein) [inline c2cgeoportal/views/tinyowsproxy.py]: > More readable syntax ...

Readable syntax, yes. But it's not doing the same thing. If you want to check if there is a non-empty string in the settings, IMHO `self.settings.get("tinyows_host", "") != ""` is an elegant way to do so.

Comment (sbrunner) [inline c2cgeoportal/views/tinyowsproxy.py]: I say again:
1. We don't need the second test
2. It's not elegant

Comment (tsauerwein) [inline c2cgeoportal/views/tinyowsproxy.py]: > I say again:
1. We don't need the second test

Thanks for repeating this, again. But you are again missing an argument why it's not needed. Like I said before, I am testing for a non-empty string, because it doesn't make sense to set an empty string as host name.

But hey, because it's only used in debug mode, I made the change. For you. :heart: 

> 2. It's not elegant

This is your opinion. I am happy to learn a more elegant way to test if there is a non-empty string for a key in the settings.

**Comment by sbrunner** [comment]
Many thanks @tsauerwein, If yous think that's ready you can merge :-)

**Comment by tsauerwein** [comment]
Thanks for the review.
